### DISCUSSION TITLE
Getting "This is not a web-friendly file type" when trying to bundle micromark with esbuild and use in a browser

### DISCUSSION THREAD

**Discussion by v1rtl** [micromark/micromark#66] [Q&A] [answered]
Hello, I'm the author of [hypermdx](https://github.com/talentlessguy/hypermdx) that uses remark-parse that depends on micromark. I tried import hypermdx in a browser:

```js
import hypermdx from 'https://cdn.skypack.dev/hypermdx';
```

 and came across with this error:

```
Uncaught Error: This is not a web-friendly file type and cannot be imported. (Imported by "parse-entities").
```

after further investigation I found out that the issue is caused by micromark.

attempting to import `micromark` results in the same error:

```js
import micromark from 'https://cdn.skypack.dev/micromark';
```

I looked in the "Source" tab in Firefox and this is what I saw:

```
/*
 * This is not a web-friendly file type and cannot be imported. (Imported by "parse-entities").
 *
 * The package "parse-entities" is attempting to import "character-entities",
 * but this is not a JavaScript module.
 *
 * How to fix:
 *   - Share your feedback if you'd like to see non-JS import support added to Skypack: https://github.com/skypackjs/skypack-cdn/issues/107
 *   - Use https://skypack.dev/ to find another web-friendly alternative to find another package.
 */

throw new Error("This is not a web-friendly file type and cannot be imported. (Imported by \"parse-entities\").");
export default null;
```

Weirdly enough, both `character-entities` and `parse-entities` import fine. But something happens in micromark that it breaks...

my question is, is it a micromark, character-entities or a skypack issue? I also tried esbuild.vercel.app (and bundling with esbuild locally) and it resulted in the same error.

## Possible solution

I'm still not sure what's causing that but *possibly* adding `exports` or `"module"` field to https://github.com/wooorm/character-entities/blob/main/package.json could help but I don't know if it will break anything or not.

**Comment by wooorm** [comment, marked as answer]
- This seems like a skypack error to me: millions of users are using micromark and parse-entities fine. For what its worth, the skypack authors are *also* using micromark in several projects. So it might be solved by configuration: to allow JSON to be imported
- The `module` field is non-standard. It could be that skypack depends on it, but it shouldn’t
- Note that I’m in the process of updating everything to ESM. The `character-entities` and `parse-entities` you see have already been updated, but `micromark` hasn’t, so be wary of that you might see code which isn’t used yet

**Reply by v1rtl** [reply]
yep looks like Skypack can't resolve JSONs yet (and btw JSON imports aren't standardized yet (see: https://github.com/tc39/proposal-json-modules))

then I'll just be waiting for when micromark will work with ESM

thanks for explaining the issue to me :D

**Reply by wooorm** [reply]
👍

> then I'll just be waiting for when micromark will work with ESM

Soon hopefully. I’ve ported about 200 projects already, and currently on micromark. The code here is updated, but I want to get some other major breaking features in (which are a bit complex). Then onto the rest!

**Reply by v1rtl** [reply]
Little update, I bumped all the deps to 10.x (e.g. ESM):

```
"hast-to-hyperscript": "^10.0.1",
"mdast-util-to-hast": "^11.2.0",
"remark-parse": "^10.0.0",
"unified": "^10.1.0"
```

and tried to import with Skypack, now getting this error in codepen:

```
"[Package Error] 'micromark@v3.0.2' could not be built. (Imported by 'mdast-util-from-markdown').
[1/5] Verifying package is valid…
[2/5] Installing dependencies from npm…
[3/5] Building package using esinstall…
Running esinstall...
@rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
ENOENT: no such file or directory, open 'micromark/stream.js'
ENOENT: no such file or directory, open '/tmp/cdn/_lftzeE15km3vT171vRWb/node_modules/micromark/stream.js'"
```

**Reply by wooorm** [reply]
are you using `stream` in your bundle?

**Reply by v1rtl** [reply]
@wooorm I don't think so?

I don't use micromark directly here:

https://github.com/talentlessguy/hypermdx/blob/master/src/mdtoh.ts

**Reply by v1rtl** [reply]
another update: importing micromark directly produces the same error:

```js
import micromark from 'https://cdn.skypack.dev/micromark';
```

```
"[Package Error] 'micromark@v3.0.2' could not be built. 
[1/5] Verifying package is valid…
[2/5] Installing dependencies from npm…
[3/5] Building package using esinstall…
Running esinstall...
@rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
ENOENT: no such file or directory, open 'micromark/stream.js'
ENOENT: no such file or directory, open '/tmp/cdn/_lftzeE15km3vT171vRWb/node_modules/micromark/stream.js'"
```

**Reply by wooorm** [reply]
Could you try forcing v3.0.1 or v3.0.0? Maybe the publish went wrong and `stream` is missing?

**Reply by wooorm** [reply]
Hmm they all don‘t include `stream`: https://unpkg.com/browse/micromark@3.0.1/

**Reply by v1rtl** [reply]
@wooorm same for 3.0.1:

```
[Package Error] "micromark@v3.0.1" could not be built. 
[1/5] Verifying package is valid…
[2/5] Installing dependencies from npm…
[3/5] Building package using esinstall…
Running esinstall...
@rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
ENOENT: no such file or directory, open 'micromark/stream.js'
ENOENT: no such file or directory, open '/tmp/cdn/_GYDAmR3172WSBN0BIhAd/node_modules/micromark/stream.js'
```

and for 3.0.0:

```
[Package Error] "micromark@v3.0.0" could not be built. 
[1/5] Verifying package is valid…
[2/5] Installing dependencies from npm…
[3/5] Building package using esinstall…
Running esinstall...
@rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
ENOENT: no such file or directory, open 'micromark/stream.js'
ENOENT: no such file or directory, open '/tmp/cdn/_vz4srgRuWfnO0KQCg9qK/node_modules/micromark/stream.js'
```

**Reply by wooorm** [reply]
working on it!

**Reply by wooorm** [reply]
should be solved: https://github.com/micromark/micromark/releases/tag/3.0.3

**Reply by v1rtl** [reply]
@wooorm yay it works! now some libs have to update micromark

thank u !

**Reply by wooorm** [reply]
which ones? remark plugins? if so, working on it!

**Reply by v1rtl** [reply]
@wooorm yup i meant them
